File: tp03/ListaSequencial.java
class Jogador
{
    private int id;
    private String nome;
    private int altura;
    private int peso;
    private String universidade;
    private int anoNascimento;
    private String cidadeNascimento;
    private String estadoNascimento;
// Construtor sem parâmetros
    public Jogador()
    {

    }
// Construtor com parâmetros para todos atributos
    public Jogador(int id, String nome, int altura, int peso, String universidade, int anoNascimento, String cidadeNascimento, String estadoNascimento)
    {
         this.id = id;
	 this.nome = nome;
	 this.altura = altura;
	 this.peso = peso;
	 this.universidade = universidade;
	 this.anoNascimento = anoNascimento;
	 this.cidadeNascimento = cidadeNascimento;
	 this.estadoNascimento = estadoNascimento;
    }
// Metódos setters
    void setId(int id)
    {
	    this.id =id;
    }
    void setNome(String nome)
    {
	    this.nome= nome;
    }
    void setAltura(int altura)
    {
	    this.altura = altura;
    }
    void setPeso(int peso)
    {
	    this.peso = peso;
    }
    void setUniversidade(String universidade)
    {
	    this.universidade = universidade;
    }
    void setAnoNascimento(int anoNascimento)
    {
	    this.anoNascimento = anoNascimento;
    }
    void setCidadeNascimento(String cidadeNascimento)
    {
	    this.cidadeNascimento = cidadeNascimento;
    } 
    void setEstadoNascimento(String estadoNascimento)
    {
	    this.estadoNascimento = estadoNascimento;
    }
// Metódos getters
    int getId()
    {
	    return id;
    }
    String getNome()
    {
	    return nome;
    }
    int getAltura()
    {
	    return altura;
    }
    int getPeso()
    {
	    return peso;
    }
    String getUniversidade()
    {
	    return universidade;
    }
    int getAnoNascimento()
    {
	    return anoNascimento;
    }
    String getCidadeNascimento()
    {
	    return cidadeNascimento;
    }
    String getEstadoNascimento()
    {
	    return estadoNascimento;
    }
//Metodo para encontrar o jogador com o id inserido
    static Jogador PesquisaId(int x)
    {
	    Jogador[] temp = ler();
	    for(int i = 0; i<3922; i++)
	    {
		    if(temp[i].getId()==x)
		    {
			    return temp[i];
		    }
	    }
	    return temp[0];
    }
// Método para ler o jogador a partir do ID do parâmetro
    static Jogador[] ler()
    {
	    Arq.openRead("/tmp/players.csv");
	    Jogador[] jogador= new Jogador[3922];
	    int i = 0;
	    Arq.readLine();
	    while(Arq.hasNext())
	    {
            jogador[i] = new Jogador(); 
	    String jog = Arq.readLine();
	    String[] temp = new String[8];
            temp= jog.split(",", 8);
	    jogador[i].setId(Integer.parseInt(temp[0]));
            jogador[i].setNome(temp[1]);
	    jogador[i].setAltura(Integer.parseInt(temp[2]));
	    jogador[i].setPeso(Integer.parseInt(temp[3]));
	    jogador[i].setUniversidade(temp[4]);
	    jogador[i].setAnoNascimento(Integer.parseInt(temp[5]));
	    jogador[i].setCidadeNascimento(temp[6]);
	    jogador[i].setEstadoNascimento(temp[7]);

// Ifs para lidar com campos vazios
	    if(temp[4].isEmpty())
	           jogador[i].setUniversidade("nao informado");  
	    if(temp[6].isEmpty())
		   jogador[i].setCidadeNascimento("nao informado");
	    if(temp[7].isEmpty())
		   jogador[i].setEstadoNascimento("nao informado");
	    i++;
            }
	    Arq.close();
	    return jogador;


    }
// Metodo para clonar um jogador
   static Jogador clonej(Jogador jogador) {
	    String temp;
	    Jogador clone = new Jogador();
	    temp =  Integer.toString(jogador.getId());
	    clone.setId(Integer.parseInt(temp));
	    temp = jogador.getNome();
	    clone.setNome(temp);
	    temp = Integer.toString(jogador.getAltura());
	    clone.setAltura(Integer.parseInt(temp));
	    temp = Integer.toString(jogador.getPeso());
	    clone.setPeso(Integer.parseInt(temp));
	    temp = jogador.getUniversidade();
	    clone.setUniversidade(temp);
	    temp = Integer.toString(jogador.getAnoNascimento());
	    clone.setAnoNascimento(Integer.parseInt(temp));
	    temp = jogador.getCidadeNascimento();
	    clone.setCidadeNascimento(temp);
	    temp = jogador.getEstadoNascimento();
	    clone.setEstadoNascimento(temp);
	    return clone;
    }
}
class ListaSequencial extends Jogador
{
	private Jogador[] array;
	private int n;

  public ListaSequencial()
  {
	  this(0);
  }
  public ListaSequencial(int tamanho)
  {
	  array = new Jogador[tamanho];
	  n = 0;
  }
  void inserirInicio(Jogador jogador) throws Exception{
  {

	if(n>= array.length)
	{
		throw new Exception("Erro ao inserir");
	}
	for(int i = n; i>0 ; i--)
	{
		array[i] = array[i-1];
	}
	array[0] = jogador;
	n++;
  }
}
  void inserir(Jogador jogador, int pos) throws Exception{
  {
	  if(pos<0 || pos>array.length)
	  {
		  throw new Exception("Posição Invalida");
	  }
	  for(int i =n; i>pos; i--)
	  {
		  array[i] = array[i-1];
	  }
	  array[pos] = jogador;
	  n++;
  }
}
  void inserirFim(Jogador jogador) throws Exception{
  {
	  if(n>= array.length)
	  {
		  throw new Exception("Array cheio");
	  }
	  array[n] = jogador;
	  n++;
  }
}
  Jogador removerInicio() throws Exception{
  {

	  if(n == 0)
	  {
		  throw new Exception("Array vazio");
	  }
	  Jogador resp = array[0];
	  for(int i =0; i<n-1; i++)
	  {
		  array[i] = array[i+1];
	  }
	  n--;
	  return resp;
  }
}
  Jogador remover(int pos) throws Exception{
  {
	  if(pos<0 || pos >=n || n==0)
	  {
		  throw new Exception("Nao foi possivel remover");
	  }
	  Jogador resp = array[pos];
	  for(int i = pos; i<n-1; i++)
	  {
		  array[i] = array[i+1];
	  }
	  n--;
	  return resp;
  }
}
  Jogador removerFim() throws Exception{
  {
	  if( n== 0)
	  {
		  throw new Exception("Array vazio");
	  }
	  return array[--n];
  }
  }
  public void preencher(Jogador[] tmp)
	{

		for(int i =0; i<n; i++)
		{
			n++;
			array[i] = clonej(tmp[i]);
		}
             System.out.println(n);
	}
  void mostrar()
{
	for(int i =0; i<n; i++)
	{
	System.out.print("["+i+"]");
	    System.out.print(" ## ");
	    System.out.print(array[i].getNome());
	    System.out.print(" ## ");
	    System.out.print(array[i].getAltura());
	    System.out.print(" ## ");
	    System.out.print(array[i].getPeso());
	    System.out.print(" ## ");
	    System.out.print(array[i].getAnoNascimento());
	    System.out.print(" ## ");
	    System.out.print(array[i].getUniversidade());
	    System.out.print(" ## ");
	    System.out.print(array[i].getCidadeNascimento());
	    System.out.print(" ## ");
	    System.out.print(array[i].getEstadoNascimento());
	    System.out.println(" ## ");
	}
}
  
  public static void main(String args[]) throws Exception {
  {
	    String temp;
	   Jogador[] a = ler();
	   Jogador[] b = new Jogador[3922];
	   int i = 0;
	   ListaSequencial ls = new ListaSequencial(3923);
	    while(true)
	    {
		   temp =  MyIO.readLine();
		   if(temp.equals("FIM"))
				   break;
                   else if(Integer.parseInt(temp) <3922)
		   {
			   ls.inserirFim(PesquisaId(Integer.parseInt(temp)));
			   i++;
		   }
              }
	    int num = MyIO.readInt();
		    for(int j = 0; j<num ;j++)
		    {
			    String acao = MyIO.readString();
			    if(acao.equals("II"))
			    {
			       int id = MyIO.readInt();
                               ls.inserirInicio(PesquisaId(id));
			    }
			    else if(acao.equals("I"))
			    {
			      int id = MyIO.readInt();
			      ls.inserirFim(PesquisaId(id));

			    }
			    else if(acao.equals("I*"))
			    {
			      int pos = MyIO.readInt();
			      int id = MyIO.readInt();
			      ls.inserir(PesquisaId(id), pos);
			    }
			    else if(acao.equals("R"))
			    {
		              Jogador removido = ls.removerFim();
			      System.out.println("(R) "+ removido.getNome());

			    }
			    else if(acao.equals("RI"))
			    {
			      Jogador removido = ls.removerInicio();
			      System.out.println("(R) "+removido.getNome());

			    }
			    else
			    {
			      int pos = MyIO.readInt();
			      Jogador removido = ls.remover(pos);
			      System.out.println("(R) "+removido.getNome());

			    }
			   
		    }
		    ls.mostrar();

  }
  }
}

File: 2º Período/tp03/ListaSequencial.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>

#define MAXTAM 3923
typedef struct Jogador
{
		int id;
		char nome[100];
		int altura;
		int peso;
		char universidade[100];
		int anoNascimento;
		char cidadeNascimento[100];
		char estadoNascimento[100];
} Jogador;
 typedef struct Lista
{
	int tamanho;
	struct Jogador array[MAXTAM];
}Lista;
 Lista NovaLista()
{
   Lista tmp; 
   tmp.tamanho = 0;
   return tmp;
}

 int getId(Jogador* x)
 {
	return x->id;
 }
 char* getNome(Jogador* x )
{
	return x->nome;
}
 int getAltura(Jogador* x)
{
	return x->altura;
}
 int getPeso(Jogador* x)
{
	return x->peso;
}
 char* getUniversidade(Jogador* x)
{
	return x->universidade;
}
 int getAnoNascimento(Jogador* x)
{
	return x->anoNascimento;
}
 char* getCidadeNascimento(Jogador* x)
{
	return x->cidadeNascimento;
}
 char* getEstadoNascimento(Jogador* x)
{
	return x->estadoNascimento;
}
void setId(Jogador *x, int novoId)
{
	x->id = novoId;
}
void setNome(Jogador* x, char* nome)
{
	strcpy(x->nome, nome);
}
void setAltura(Jogador* x, int altura)
{
	x->altura = altura;
}
void setPeso(Jogador* x, int peso)
{
	x->peso = peso;
}
void setUniversidade(Jogador *x, char* universidade)
{
	strcpy(x->universidade, universidade);
}
void setAnoNascimento(Jogador* x, int anoNascimento)
{
	x->anoNascimento = anoNascimento;
}
void setCidadeNascimento(Jogador* x, char* cidadeNascimento)
{
	strcpy(x->cidadeNascimento, cidadeNascimento);
}
void setEstadoNascimento(Jogador* x, char* estadoNascimento)
{
	strcpy(x->estadoNascimento, estadoNascimento);
}
//metodo para criar um clone do jogador
Jogador clone(Jogador *x)
{
	Jogador clone;
	clone.id = getId(x);
	strcpy(clone.nome, getNome(x));
	clone.altura = getAltura(x);
        clone.peso = getPeso(x);
        clone.anoNascimento = getAnoNascimento(x);
        strcpy(clone.universidade, getUniversidade(x));
	strcpy(clone.cidadeNascimento, getCidadeNascimento(x));
	strcpy(clone.estadoNascimento, getEstadoNascimento(x));

	return clone;
}
//Metodo para ler o arquivo e retornar um ponteiro para os Jogadores selecionados
Jogador* ler()
{
 FILE* arq = fopen("/tmp/players.csv","r");
 if(arq == NULL)
 {
	 printf("nao abriu");
 }
 char temp[1000];
 int cont;
 int num = 0;
 Jogador *y=malloc(sizeof(Jogador));
 Jogador *resp = malloc(3923*sizeof(Jogador));
   while(fscanf(arq, "%[^\n]", temp) != EOF)
   {
	   cont = 0; 
	   fgetc(arq);
	   char *items[8];
	   items[4]="";
	   items[6]="";
	   items[7]="";
           char *token  = strtok(temp,",");
	   for(int i = 0; token ; i++)
	   {
            items[i] = token;
            token = strtok(NULL, ",");
            cont++;
	   }
//ifs para tratar casos em que existiam informações faltantes
	   if(cont == 7 && strlen(items[7])==0)
       {
           items[7] = items[6];
       }
	   else if(cont ==5)
	   {
		   items[5] = items[4];
		   items[4] = "";
	   }
       if(atoi(items[5]) == 0)
       {
           items[6] = items [5];
       }
	   if(atoi(items[4]) != 0)
       {
           items[5] = items[4];
       }

	   setId(y, atoi(items[0]));
	   setNome(y, items[1]);
	   setAltura(y, atoi(items[2]));
	   setPeso(y, atoi(items[3]));
	   setUniversidade(y, items[4]);
	   setAnoNascimento(y, atoi(items[5]));
       setCidadeNascimento(y, items[6]);
	   setEstadoNascimento(y, items[7]);

//ifs para setar corretamente campos não informados
           if(strlen(items[4])==0 || atoi(getUniversidade(y))>0)
	   {
	           setUniversidade(y, "nao informado");
	   }
	     if(strlen(items[6])==0)
	   {
		  setCidadeNascimento(y, "nao informado");
	   }
	     if(strlen(items[7])==0)
	   {
		   setEstadoNascimento(y, "nao informado");
	   }
	     if(getAnoNascimento(y)==0)
         {
             setAnoNascimento(y, atoi(items[4]));
         }
         
//comparar o input do usuario com o que esta sendo lid
	 resp[num] =clone(y);
	 num++;
   }
   fclose(arq);
   return resp;
}

//função para imprimir o jogador selecionado
void imprimir(Lista *l)
{
	for(int i =0; i<l->tamanho; i++)
	{
	printf("[");
	printf("%d]", i);
	printf(" ## ");
	printf("%s", getNome(&l->array[i]));
	printf(" ## ");
	printf("%d", getAltura(&l->array[i]));
	printf(" ## ");
	printf("%d", getPeso(&l->array[i]));
	printf(" ## ");
	printf("%d", getAnoNascimento(&l->array[i]));
	printf(" ## "); 
	printf("%s", getUniversidade(&l->array[i]));
	printf(" ## "); 
	printf("%s", getCidadeNascimento(&l->array[i]));
	printf(" ## ");
	printf("%s ##", getEstadoNascimento(&l->array[i]));
	printf("\n");
	}
}
//metodo para trocar dois jogadores de posição no array
void swap(Jogador *i, Jogador *j)
{
	Jogador tmp = *i;
	*i = *j;
	*j = tmp;
	
}
//metodo para comparar jogadores com o atributo primário sendo o EstadoNascimento e o secundário o nome
 int compare2(Jogador *x, Jogador *y) 
{
    if (getAltura(x) != getAltura(y)) {
        return (getAltura(x) > getAltura(y));
    } else {
		return strcmp(getNome(x), getNome(y))>0;
    }
}
void inserirInicio(Jogador x, Lista *l)
{
	if(l->tamanho >= MAXTAM)
	{ 
            printf("erro");
	    return;
	}
	for(int i = l->tamanho; i>0; i--)
	{
		l->array[i] = l->array[i-1];

	}
	l->array[0] = x;
	l->tamanho++;
	
}
void inserirFim(Jogador x, Lista *l)
{
	if(l->tamanho >= MAXTAM)
	{
		printf("erro");
		return;
	}
		l->array[l->tamanho] = x;
		l->tamanho++;
}
void inserir(int pos, Jogador x, Lista *l)
{
	if(pos <0 || pos> l->tamanho || l->tamanho >= MAXTAM)
		return;
	else
	{
		for(int i = l->tamanho; i>pos; i--)
		{
			
			l->array[i] = l->array[i-1];
		}
		l->array[pos] = x;
		l->tamanho++;
	}

}
Jogador removerInicio(Lista *l)
{
       if(l->tamanho == 0)
       {
	       printf("Array vazio");
	       exit(1);
       }
       Jogador removido = l->array[0];	 
       for(int i =0; i<l->tamanho; i++)
       {
	       l->array[i] = l->array[i+1];
       }
       l->tamanho--;
       return removido;
}
Jogador removerFim(Lista *l)
{
	if(l->tamanho == 0)
	{
		printf("Array vazio");
		exit(1);
	}

	return l->array[--l->tamanho];

}
Jogador remover(int pos, Lista *l)
{
	Jogador removido;
	if(pos<0 || pos >= l->tamanho-1)
	{
		printf("Posição invalida");
		exit(1);
	}

	else if(pos == 0)
	{
		removido = removerInicio(l);
	}
	else if(pos == l->tamanho-1)
	{
		removido = removerFim(l);
	}
	else
	{
		removido = l->array[pos];
		for(int i = pos; i< l->tamanho-1; i++)
		{
			l->array[i] = l->array[i+1];
		}
		l->tamanho--;

	}
	return removido;
}

int main()
{ 
//criação de variavéis e alocação de memória
  char temp[100];
  int i =0;
  Lista l = NovaLista();
  Jogador *y = malloc(3923*sizeof(Jogador));
  y = ler();
  //while para leitura do input ate ser digitado FIM
  while(1)
  {
  scanf("%[^\n]", temp);
  char ch = getchar();
   if(strcmp(temp, "FIM")== 0)
   {
		  break; 
   }
   int num = atoi(temp);
   inserirFim(y[num+1], &l); 
  }
  int qnt;
  scanf("%d", &qnt);
  char ch = getchar();
 for(int j = 0; j<qnt; j++)
  {
	  char acao[2];
	  scanf("%s", acao);
	 if(strcmp(acao, "II")== 0)
		 {
		             int id;
			     scanf("%d", &id);
			     ch =getchar();
                             inserirInicio(y[id+1], &l);
		} 
	 else if(strcmp(acao, "IF")== 0)
	 {
		 int id;
		 scanf("%d", &id);
		 ch = getchar();
		 inserirFim(y[id+1], &l);
	 }
	 else if(strcmp(acao, "I*")==0)
	 {
		 int id;
		 int pos;
		 scanf("%d %d", &pos, &id);
		 inserir(pos, y[id+1], &l);
	 }
	 else if(strcmp(acao, "RI")==0)
	 {
		 Jogador removido = removerInicio(&l);
		 printf("(R) %s\n", getNome(&removido));
	 }
	 else if(strcmp(acao, "RF")==0)
	 {
	         Jogador removido = removerFim(&l);
		 printf("(R) %s\n", getNome(&removido));
	 }
	 else
	 {
		 int pos;
		 scanf("%d", &pos);
		 ch = getchar();
		 Jogador removido = remover(pos, &l);
		 printf("(R) %s\n", getNome(&removido));
	 }
  }
  imprimir(&l);
  free(y);
  return 0;
}
